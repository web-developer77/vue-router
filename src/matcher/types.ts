import { RouteRecordMultipleViews, RouteRecordRedirect } from '../types'

interface RouteRecordRedirectNormalized {
  path: RouteRecordRedirect['path']
  name: RouteRecordRedirect['name']
  redirect: RouteRecordRedirect['redirect']
  meta: RouteRecordRedirect['meta']
  beforeEnter: RouteRecordRedirect['beforeEnter']
}
interface RouteRecordViewNormalized {
  path: RouteRecordMultipleViews['path']
  name: RouteRecordMultipleViews['name']
  components: RouteRecordMultipleViews['components']
  children: RouteRecordMultipleViews['children']
  meta: RouteRecordMultipleViews['meta']
  beforeEnter: RouteRecordMultipleViews['beforeEnter']
}
// normalize component/components into components
export type RouteRecordNormalized =
  | RouteRecordRedirectNormalized
  | RouteRecordViewNormalized
