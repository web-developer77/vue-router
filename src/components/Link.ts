import { defineComponent, h, PropType, inject } from '@vue/runtime-core'
import { computed, reactive, isRef, Ref } from '@vue/reactivity'
import { RouteLocation } from '../types'

export function useLink(to: Ref<RouteLocation> | RouteLocation) {
  const router = inject('router')

  const route = computed(() => router.resolve(isRef(to) ? to.value : to))
  const href = computed(() => router.createHref(route.value))
  const isActive = computed<boolean>(
    () => router.currentRoute.value.path.indexOf(route.value.path) === 0
  )

  // TODO: handle replace prop

  function navigate(e: MouseEvent) {
    // TODO: handle navigate with empty parameters for scoped slot and composition api
    if (guardEvent(e)) router.push(route.value)
  }

  return {
    route,
    href,
    isActive,
    navigate,
  }
}

const Link = defineComponent({
  name: 'RouterLink',
  props: {
    to: {
      type: [String, Object] as PropType<RouteLocation>,
      required: true,
    },
  },

  setup(props, { slots, attrs }) {
    const { route, isActive, href, navigate } = useLink(props.to)

    const elClass = computed(() => ({
      'router-link-active': isActive.value,
    }))

    // TODO: exact active classes
    // TODO: handle replace prop

    return () => {
      return h(
        'a',
        {
          class: elClass.value,
          onClick: navigate,
          href: href.value,
          ...attrs,
        },
        slots.default(reactive({ route, href, isActive }))
      )
    }
  },
})

function guardEvent(e: MouseEvent) {
  // don't redirect with control keys
  if (e.metaKey || e.altKey || e.ctrlKey || e.shiftKey) return
  // don't redirect when preventDefault called
  if (e.defaultPrevented) return
  // don't redirect on right click
  if (e.button !== undefined && e.button !== 0) return
  // don't redirect if `target="_blank"`
  // @ts-ignore getAttribute does exist
  if (e.currentTarget && e.currentTarget.getAttribute) {
    // @ts-ignore getAttribute exists
    const target = e.currentTarget.getAttribute('target')
    if (/\b_blank\b/i.test(target)) return
  }
  // this may be a Weex event which doesn't have this method
  if (e.preventDefault) e.preventDefault()

  return true
}

export default Link
