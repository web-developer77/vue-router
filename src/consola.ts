// Own simplified consola version
// Consola doesn't work on IE9

type LogCommand = 'info' | 'log' | 'error' | 'warn'
const logs: LogCommand[] = ['info', 'log', 'error', 'warn']

export default {
  tag: '',
  withTag(tag: string) {
    const cs = Object.create(this)
    cs.tag = tag
    return cs
  },

  info(...args: any[]) {
    if (this.tag) console.info(`[${this.tag}]`, ...args)
    else console.info(...args)
  },

  mockTypes(spyCreator: () => any): void {
    for (const log of logs) {
      // @ts-ignore
      this[log] = spyCreator()
    }
  },
}
