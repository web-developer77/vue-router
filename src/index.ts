import { Router } from './router'
import { HTML5History } from './history/html5'

export { Router, HTML5History }
