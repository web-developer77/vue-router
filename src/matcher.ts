import pathToRegexp from 'path-to-regexp'
import {
  RouteRecord,
  RouteParams,
  MatcherLocation,
  MatcherLocationNormalized,
  MatcherLocationRedirect,
  // TODO: add it to matched
  // MatchedRouteRecord,
} from './types/index'
import { NoRouteMatchError, InvalidRouteMatch } from './errors'

type NormalizedRouteRecord = Exclude<RouteRecord, { component: any }> // normalize component/components into components

interface RouteMatcher {
  re: RegExp
  resolve: (params?: RouteParams) => string
  record: NormalizedRouteRecord
  parent: RouteMatcher | void
  // TODO: children so they can be removed
  // children: RouteMatcher[]
  keys: string[]
  score: number
}

/**
 * Normalizes a RouteRecord into a MatchedRouteRecord. Creates a copy
 * @param record
 * @returns the normalized version
 */
export function normalizeRecord(
  record: Readonly<RouteRecord>
): NormalizedRouteRecord {
  if ('component' in record) {
    const { component, ...rest } = record
    // @ts-ignore I could do it type safe by copying again rest:
    // return {
    //   ...rest,
    //   components: { default: component }
    // }
    // but it's slower
    rest.components = { default: component }
    return rest as NormalizedRouteRecord
  }

  // otherwise just create a copy
  return { ...record }
}

enum PathScore {
  Segment = 4, // /a-segment
  SubSegment = 2, // /multiple-:things-in-one-:segment
  Static = 3, // /static
  Dynamic = 2, // /:someId
  DynamicCustomRegexp = 2.5, // /:someId(\\d+)
  Wildcard = -1, // /:namedWildcard(.*)
  SubWildcard = 1, // Wildcard as a subsegment
  Repeatable = -0.5, // /:w+ or /:w*
  Optional = -4, // /:w? or /:w*
  SubOptional = -0.1, // optional inside a subsegment /a-:w? or /a-:w*
  Root = 1, // just /
}

// allows to check if the user provided a custom regexp
const isDefaultPathRegExpRE = /^\[\^[^\]]+\]\+\?$/

export function createRouteMatcher(
  record: Readonly<NormalizedRouteRecord>,
  parent: RouteMatcher | void,
  options: pathToRegexp.RegExpOptions
): RouteMatcher {
  const keys: pathToRegexp.Key[] = []
  // options only use `delimiter`
  const tokens = pathToRegexp.parse(record.path, options)
  const re = pathToRegexp.tokensToRegExp(tokens, keys, options)
  // we pass a copy because later on we are modifying the original token array
  // to compute the score of routes
  const resolve = pathToRegexp.tokensToFunction([...tokens])

  let score = 0

  // console.log(tokens)
  // console.log('--- GROUPING ---')

  // special case for root path
  if (tokens.length === 1 && tokens[0] === '/') {
    score = 5
  } else {
    // allows us to group tokens into one single segment
    // it will point to the first token of the current group
    let currentSegment = 0
    // we group them in arrays to process them later
    const groups: Array<pathToRegexp.Token[]> = []
    // we skip the first element as it must be part of the first group
    const token = tokens[0]
    if (typeof token === 'string') {
      // TODO: refactor code in loop (right now it is duplicated)
      // we still need to check for / inside the string
      // remove the empty string because of the leading slash
      const sections = token.split('/').slice(1)
      if (sections.length > 1) {
        // the last one is going to replace currentSegment
        const last = sections.pop() as string // ts complains but length >= 2
        // we need to finalize previous group but we cannot use current entry
        // here we are sure that currentSegment < i because the token doesn't start with /
        // assert(currentSegment < i)
        const first = sections.shift() as string // ts complains but length >= 2
        // so we cut until the current segment and add the first section of current token as well
        groups.push(tokens.slice(currentSegment, 0).concat('/' + first))
        // equivalent to
        // groups.push(['/' + first])

        // we add the remaining sections as static groups
        for (const section of sections) {
          groups.push(['/' + section])
        }

        // we replace current entry with the last section and reset current segment
        tokens[0] = '/' + last
        // currentSegment = 0
      }
    }

    for (let i = 1; i < tokens.length; i++) {
      const token = tokens[i]
      if (typeof token === 'string') {
        if (token.charAt(0) === '/') {
          // finalize previous group and start a new one
          groups.push(tokens.slice(currentSegment, i))
          currentSegment = i
        } else {
          // we still need to check for / inside the string
          const sections = token.split('/')
          if (sections.length > 1) {
            // the last one is going to replace currentSegment
            const last = sections.pop() as string // ts complains but length >= 2
            // we need to finalize previous group but we cannot use current entry
            // here we are sure that currentSegment < i because the token doesn't start with /
            // assert(currentSegment < i)
            const first = sections.shift() as string // ts complains but length >= 2
            // so we cut until the current segment and add the first section of current token as well
            groups.push(tokens.slice(currentSegment, i).concat(first))

            // we add the remaining sections as static groups
            for (const section of sections) {
              groups.push(['/' + section])
            }

            // we replace current entry with the last section and reset current segment
            tokens[i] = '/' + last
            currentSegment = i
          }
        }
      } else {
        if (token.prefix.charAt(0) === '/') {
          // finalize previous group and start a new one
          groups.push(tokens.slice(currentSegment, i))
          currentSegment = i
        }
      }
    }

    // add the remaining segment as one group
    // TODO: refactor the handling of ending with static like /:a/b/c
    if (currentSegment < tokens.length) {
      let token: pathToRegexp.Token
      if (
        tokens.length - currentSegment === 1 &&
        typeof (token = tokens[tokens.length - 1]) === 'string'
      ) {
        // the remaining group is a single string, so it must start with a leading /
        const sections = token.split('/').slice(1)
        // we add the remaining sections as static groups
        for (const section of sections) {
          groups.push(['/' + section])
        }
      } else {
        groups.push(tokens.slice(currentSegment))
      }
    }

    const scoreForSegment = function scoreForSegment(
      group: pathToRegexp.Token
    ): number {
      let score = PathScore.Segment
      if (typeof group === 'string') {
        score += group === '/' ? PathScore.Root : PathScore.Static
      } else {
        score +=
          group.pattern === '.*'
            ? PathScore.Wildcard
            : isDefaultPathRegExpRE.test(group.pattern)
            ? PathScore.Dynamic
            : PathScore.DynamicCustomRegexp
        score +=
          +group.optional * PathScore.Optional +
          +group.repeat * PathScore.Repeatable
        if (typeof group.name === 'number') throw new Error('Name your param')
        // keys.push(group.name)
      }
      return score
    }

    const scoreForSubSegment = function scoreForSubSegment(
      group: pathToRegexp.Token
    ): number {
      let score = 0
      if (typeof group === 'string') {
        // in a sub segment, it doesn't matter if it's root or not
        score += PathScore.Static
      } else {
        score +=
          group.pattern === '.*'
            ? PathScore.SubWildcard
            : isDefaultPathRegExpRE.test(group.pattern)
            ? PathScore.Dynamic
            : PathScore.DynamicCustomRegexp
        score += +group.optional * PathScore.SubOptional
        if (typeof group.name === 'number') throw new Error('Name your param')
        // keys.push(group.name)
      }
      return score
    }

    for (const group of groups) {
      // console.log(group)
      if (group.length === 1) {
        score += scoreForSegment(group[0])
      } else {
        score += PathScore.Segment + PathScore.SubSegment
        let multiplier = 1 / 10
        for (let i = 0; i < group.length; i++) {
          score += scoreForSubSegment(group[i]) * multiplier
          multiplier /= 10
        }
      }
      // segments.push('/' + section)
    }

    // console.log(record.path, { score })
    // console.log('____'.repeat(20))
  }

  // create the object before hand so it can be passed to children
  return {
    parent,
    record: record,
    re,
    // TODO: handle numbers differently. Maybe take the max one and say there are x unnamed keys
    keys: keys.map(key => String(key.name)),
    resolve,
    score,
  }
}

export class RouterMatcher {
  private matchers: RouteMatcher[] = []

  constructor(routes: RouteRecord[]) {
    for (const route of routes) {
      this.addRouteRecord(route)
    }
  }

  private addRouteRecord(
    record: Readonly<RouteRecord>,
    parent?: RouteMatcher
  ): void {
    const options: pathToRegexp.RegExpOptions = {}

    const recordCopy = normalizeRecord(record)

    if (parent) {
      // if the child isn't an absolute route
      if (record.path[0] !== '/') {
        let path = parent.record.path
        // only add the / delimiter if the child path isn't empty
        if (recordCopy.path) path += '/'
        path += record.path
        recordCopy.path = path
      }
    }

    // create the object before hand so it can be passed to children
    const matcher = createRouteMatcher(recordCopy, parent, options)

    if ('children' in record && record.children) {
      for (const childRecord of record.children) {
        this.addRouteRecord(childRecord, matcher)
      }
    }

    this.insertMatcher(matcher)
  }

  private insertMatcher(matcher: RouteMatcher) {
    let i = 0
    while (i < this.matchers.length && matcher.score <= this.matchers[i].score)
      i++
    this.matchers.splice(i, 0, matcher)
  }

  /**
   * Resolve a location without doing redirections so it can be used for anchors
   */
  resolveAsPath() {}

  /**
   * Transforms a MatcherLocation object into a normalized location
   * @param location MatcherLocation to resolve to a url
   */
  resolve(
    location: Readonly<MatcherLocation>,
    currentLocation: Readonly<MatcherLocationNormalized>
  ): MatcherLocationNormalized | MatcherLocationRedirect {
    let matcher: RouteMatcher | void
    let params: RouteParams = {}
    let path: MatcherLocationNormalized['path']
    let name: MatcherLocationNormalized['name']

    if ('path' in location) {
      matcher = this.matchers.find(m => m.re.test(location.path))

      // TODO: if no matcher, return the location with an empty matched array
      // to allow non existent matches
      // TODO: warning of unused params if provided
      if (!matcher) throw new NoRouteMatchError(currentLocation, location)

      // no need to resolve the path with the matcher as it was provided
      path = location.path
      name = matcher.record.name

      // fill params
      const result = matcher.re.exec(path)

      if (!result) {
        throw new Error(`Error parsing path "${location.path}"`)
      }

      for (let i = 0; i < matcher.keys.length; i++) {
        const key = matcher.keys[i]
        const value = result[i + 1]
        if (!value) {
          throw new Error(
            `Error parsing path "${location.path}" when looking for param "${key}"`
          )
        }
        params[key] = value
      }

      if ('redirect' in matcher.record) {
        const { redirect } = matcher.record
        return {
          redirect,
          normalizedLocation: {
            name,
            path,
            matched: [],
            params,
          },
        }
      }
      // named route
    } else if ('name' in location) {
      matcher = this.matchers.find(m => m.record.name === location.name)

      if (!matcher) throw new NoRouteMatchError(currentLocation, location)

      name = matcher.record.name
      params = location.params || currentLocation.params // TODO: normalize params
      path = matcher.resolve(params)
      // TODO: check missing params

      if ('redirect' in matcher.record) {
        const { redirect } = matcher.record
        return {
          redirect,
          normalizedLocation: {
            name,
            path,
            matched: [],
            params,
          },
        }
      }
      // location is a relative path
    } else {
      // match by name or path of current route
      matcher = currentLocation.name
        ? this.matchers.find(m => m.record.name === currentLocation.name)
        : this.matchers.find(m => m.re.test(currentLocation.path))
      if (!matcher) throw new NoRouteMatchError(currentLocation, location)
      name = matcher.record.name
      params = location.params || currentLocation.params
      path = matcher.resolve(params)
    }

    // this should never happen because it will mean that the user ended up in a route
    // that redirects but ended up not redirecting
    if ('redirect' in matcher.record) throw new InvalidRouteMatch(location)

    const matched: MatcherLocationNormalized['matched'] = [matcher.record]
    let parentMatcher: RouteMatcher | void = matcher.parent
    while (parentMatcher) {
      // reversed order so parents are at the beginning
      // TODO: should be doable by typing RouteMatcher in a different way
      if ('redirect' in parentMatcher.record) throw new Error('TODO')
      matched.unshift(parentMatcher.record)
      parentMatcher = parentMatcher.parent
    }

    return {
      name,
      path,
      params,
      matched,
    }
  }
}
