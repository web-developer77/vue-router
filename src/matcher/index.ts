import {
  RouteRecord,
  MatcherLocation,
  MatcherLocationNormalized,
  ListenerRemover,
} from '../types'
import { NoRouteMatchError } from '../errors'
import { createRouteRecordMatcher, RouteRecordMatcher } from './path-matcher'
import { RouteRecordNormalized } from './types'
import {
  PathParams,
  comparePathParserScore,
  PathParserOptions,
} from './path-parser-ranker'

interface RouterMatcher {
  addRoute: (
    record: RouteRecord,
    parent?: RouteRecordMatcher
  ) => ListenerRemover
  removeRoute: {
    (matcher: RouteRecordMatcher): void
    (name: Required<RouteRecord>['name']): void
  }
  getRoutes: () => RouteRecordMatcher[]
  getRecordMatcher: (
    name: Required<RouteRecord>['name']
  ) => RouteRecordMatcher | undefined
  resolve: (
    location: Readonly<MatcherLocation>,
    currentLocation: Readonly<MatcherLocationNormalized>
  ) => MatcherLocationNormalized
}

export function createRouterMatcher(
  routes: RouteRecord[],
  globalOptions: PathParserOptions
): RouterMatcher {
  // normalized ordered array of matchers
  const matchers: RouteRecordMatcher[] = []
  const matcherMap = new Map<string | symbol, RouteRecordMatcher>()

  function getRecordMatcher(name: string) {
    return matcherMap.get(name)
  }

  // TODO: add routes to children of parent
  function addRoute(
    record: Readonly<RouteRecord>,
    parent?: RouteRecordMatcher
  ) {
    const mainNormalizedRecord = normalizeRouteRecord(record)
    const options: PathParserOptions = { ...globalOptions, ...record.options }
    // generate an array of records to correctly handle aliases
    const normalizedRecords: RouteRecordNormalized[] = [mainNormalizedRecord]
    if ('alias' in record) {
      const aliases =
        typeof record.alias === 'string' ? [record.alias] : record.alias!
      for (const alias of aliases) {
        normalizedRecords.push({
          ...mainNormalizedRecord,
          path: alias,
          aliasOf: mainNormalizedRecord,
        })
      }
    }

    let matcher: RouteRecordMatcher

    for (const normalizedRecord of normalizedRecords) {
      let { path } = normalizedRecord
      // Build up the path for nested routes if the child isn't an absolute
      // route. Only add the / delimiter if the child path isn't empty and if the
      // parent path doesn't have a trailing slash
      if (parent && path[0] !== '/') {
        let parentPath = parent.record.path
        let connectingSlash =
          parentPath[parentPath.length - 1] === '/' ? '' : '/'
        normalizedRecord.path =
          parent.record.path + (path && connectingSlash + path)
      }

      // create the object before hand so it can be passed to children
      matcher = createRouteRecordMatcher(normalizedRecord, parent, options)

      if ('children' in record) {
        for (const childRecord of record.children!)
          addRoute(childRecord, matcher)
      }

      insertMatcher(matcher)
    }

    return () => {
      // since other matchers are aliases, they should should be removed by any of the matchers
      removeRoute(matcher)
    }
  }

  function removeRoute(matcherRef: string | RouteRecordMatcher) {
    // TODO: remove aliases (needs to keep them in the RouteRecordMatcher first)
    if (typeof matcherRef === 'string') {
      const matcher = matcherMap.get(matcherRef)
      if (matcher) {
        matcherMap.delete(matcherRef)
        matchers.splice(matchers.indexOf(matcher), 1)
        matcher.children.forEach(removeRoute)
      }
    } else {
      let index = matchers.indexOf(matcherRef)
      if (index > -1) {
        matchers.splice(index, 1)
        if (matcherRef.record.name) matcherMap.delete(matcherRef.record.name)
        matcherRef.children.forEach(removeRoute)
      }
    }
  }

  function getRoutes() {
    return matchers
  }

  function insertMatcher(matcher: RouteRecordMatcher) {
    let i = 0
    // console.log('i is', { i })
    while (
      i < matchers.length &&
      comparePathParserScore(matcher, matchers[i]) >= 0
    )
      i++
    // console.log('END i is', { i })
    // while (i < matchers.length && matcher.score <= matchers[i].score) i++
    matchers.splice(i, 0, matcher)
    // only add the original record to the name map
    if (matcher.record.name && !matcher.record.aliasOf)
      matcherMap.set(matcher.record.name, matcher)
  }

  /**
   * Resolves a location. Gives access to the route record that corresponds to the actual path as well as filling the corresponding params objects
   * @param location MatcherLocation to resolve to a url
   * @param currentLocation MatcherLocationNormalized of the current location
   */
  function resolve(
    location: Readonly<MatcherLocation>,
    currentLocation: Readonly<MatcherLocationNormalized>
  ): MatcherLocationNormalized {
    let matcher: RouteRecordMatcher | undefined
    let params: PathParams = {}
    let path: MatcherLocationNormalized['path']
    let name: MatcherLocationNormalized['name']

    if ('name' in location && location.name) {
      matcher = matcherMap.get(location.name)

      if (!matcher) throw new NoRouteMatchError(location)

      name = matcher.record.name
      // TODO: merge params with current location. Should this be done by name. I think there should be some kind of relationship between the records like children of a parent should keep parent props but not the rest
      // needs an RFC if breaking change
      params = location.params || currentLocation.params
      // throws if cannot be stringified
      path = matcher.stringify(params)
    } else if ('path' in location) {
      matcher = matchers.find(m => m.re.test(location.path))
      // matcher should have a value after the loop

      // no need to resolve the path with the matcher as it was provided
      // this also allows the user to control the encoding
      path = location.path
      if (matcher) {
        // TODO: dev warning of unused params if provided
        params = matcher.parse(location.path)!
        name = matcher.record.name
      }
      // location is a relative path
    } else {
      // match by name or path of current route
      matcher = currentLocation.name
        ? matcherMap.get(currentLocation.name)
        : matchers.find(m => m.re.test(currentLocation.path))
      if (!matcher) throw new NoRouteMatchError(location, currentLocation)
      name = matcher.record.name
      params = location.params || currentLocation.params
      path = matcher.stringify(params)
    }

    const matched: MatcherLocationNormalized['matched'] = []
    let parentMatcher: RouteRecordMatcher | void = matcher
    while (parentMatcher) {
      // reversed order so parents are at the beginning
      // const { record } = parentMatcher
      // TODO: check resolving child routes by path when parent has an alias
      matched.unshift(parentMatcher.record)
      parentMatcher = parentMatcher.parent
    }

    return {
      name,
      path,
      params,
      matched,
      meta: matcher ? matcher.record.meta : {},
    }
  }

  // add initial routes
  routes.forEach(route => addRoute(route))

  return { addRoute, resolve, removeRoute, getRoutes, getRecordMatcher }
}

/**
 * Normalizes a RouteRecord. Transforms the `redirect` option into a `beforeEnter`
 * @param record
 * @returns the normalized version
 */
export function normalizeRouteRecord(
  record: Readonly<RouteRecord>
): RouteRecordNormalized {
  let components: RouteRecordNormalized['components']
  let beforeEnter: RouteRecordNormalized['beforeEnter']
  if ('redirect' in record) {
    components = {}
    let { redirect } = record
    beforeEnter = (to, from, next) => {
      next(typeof redirect === 'function' ? redirect(to) : redirect)
    }
  } else {
    components =
      'components' in record ? record.components : { default: record.component }
    beforeEnter = record.beforeEnter
  }

  return {
    path: record.path,
    components,
    // fallback to empty array for monomorphic objects
    children: (record as any).children,
    name: record.name,
    beforeEnter,
    meta: record.meta || {},
    leaveGuards: [],
    aliasOf: undefined,
  }
}

export { PathParserOptions }
