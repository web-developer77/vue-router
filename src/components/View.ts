import {
  h,
  inject,
  provide,
  defineComponent,
  PropType,
  computed,
  Component,
  InjectionKey,
  Ref,
} from 'vue'
import { RouteRecordNormalized } from '../matcher/types'

// TODO: make it work with no symbols too for IE
export const matchedRouteKey = Symbol() as InjectionKey<
  Ref<RouteRecordNormalized>
>

export const View = defineComponent({
  name: 'RouterView',
  props: {
    name: {
      type: String as PropType<string>,
      default: 'default',
    },
  },

  setup(props, { attrs }) {
    const route = inject('route')
    const depth: number = inject('routerViewDepth', 0)
    provide('routerViewDepth', depth + 1)

    const matchedRoute = computed(() => route.value.matched[depth])
    const ViewComponent = computed<Component | undefined>(
      () => matchedRoute.value && matchedRoute.value.components[props.name]
    )

    provide(matchedRouteKey, matchedRoute)

    return () => {
      return ViewComponent.value
        ? h(ViewComponent.value as any, { ...attrs })
        : null
    }
  },
})
