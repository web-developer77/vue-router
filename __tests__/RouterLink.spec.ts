/**
 * @jest-environment jsdom
 */
import { Link as RouterLink } from '../src/components/Link'
import {
  START_LOCATION_NORMALIZED,
  RouteQueryAndHash,
  MatcherLocation,
  RouteLocationNormalized,
} from '../src/types'
import { createMemoryHistory } from '../src'
import { mount, tick } from './mount'
import { ref, markNonReactive } from 'vue'
import { RouteRecordNormalized } from '../src/matcher/types'

const records = {
  home: {} as RouteRecordNormalized,
  foo: {} as RouteRecordNormalized,
}

const locations: Record<
  string,
  {
    string: string
    normalized: RouteLocationNormalized
    toResolve?: MatcherLocation & Required<RouteQueryAndHash>
  }
> = {
  basic: {
    string: '/home',
    // toResolve: { path: '/home', fullPath: '/home', undefined, query: {}, hash: '' },
    normalized: {
      fullPath: '/home',
      path: '/home',
      params: {},
      meta: {},
      query: {},
      hash: '',
      matched: [records.home],
      redirectedFrom: undefined,
      name: undefined,
    },
  },
  foo: {
    string: '/foo',
    // toResolve: { path: '/home', fullPath: '/home', undefined, query: {}, hash: '' },
    normalized: {
      fullPath: '/foo',
      path: '/foo',
      params: {},
      meta: {},
      query: {},
      hash: '',
      matched: [records.foo],
      redirectedFrom: undefined,
      name: undefined,
    },
  },
  withQuery: {
    string: '/home?foo=a&bar=b',
    // toResolve: { path: '/home', fullPath: '/home', undefined, query: {}, hash: '' },
    normalized: {
      fullPath: '/home?foo=a&bar=b',
      path: '/home',
      params: {},
      meta: {},
      query: { foo: 'a', bar: 'b' },
      hash: '',
      matched: [records.home],
      redirectedFrom: undefined,
      name: undefined,
    },
  },
  repeatedParams2: {
    string: '/p/1/2',
    normalized: {
      fullPath: '/p/1/2',
      path: '/p/1/2',
      params: { p: ['1', '2'] },
      meta: {},
      query: {},
      hash: '',
      matched: [records.home],
      redirectedFrom: undefined,
      name: undefined,
    },
  },
  repeatedParams3: {
    string: '/p/1/2/3',
    normalized: {
      fullPath: '/p/1/2/3',
      path: '/p/1/2/3',
      params: { p: ['1', '2', '3'] },
      meta: {},
      query: {},
      hash: '',
      matched: [records.home],
      redirectedFrom: undefined,
      name: undefined,
    },
  },
}

describe('RouterLink', () => {
  function factory(
    currentLocation: RouteLocationNormalized,
    propsData: any,
    resolvedLocation: RouteLocationNormalized
  ) {
    const router = {
      history: createMemoryHistory(),
      createHref(to: RouteLocationNormalized): string {
        return this.history.base + to.fullPath
      },
      resolve: jest.fn(),
      push: jest.fn().mockResolvedValue(resolvedLocation),
      currentRoute: ref(markNonReactive(currentLocation)),
      setActiveApp: jest.fn(),
    }

    router.resolve.mockReturnValueOnce(resolvedLocation)
    const { app, el } = mount(router as any, {
      template: `<RouterLink :to="to">a link</RouterLink>`,
      components: { RouterLink } as any,
      setup() {
        return { to: propsData.to }
      },
    })

    return { app, router, el }
  }

  it('displays a link with a string prop', () => {
    const { el } = factory(
      START_LOCATION_NORMALIZED,
      { to: locations.basic.string },
      locations.basic.normalized
    )
    expect(el.querySelector('a')!.getAttribute('href')).toBe('/home')
  })

  it('can change the value', async () => {
    const to = ref(locations.basic.string)
    const { el, router } = factory(
      START_LOCATION_NORMALIZED,
      { to },
      locations.basic.normalized
    )
    expect(el.querySelector('a')!.getAttribute('href')).toBe('/home')
    router.resolve.mockReturnValueOnce(locations.foo.normalized)
    to.value = locations.foo.string
    await tick()
    expect(el.querySelector('a')!.getAttribute('href')).toBe('/foo')
  })

  it('displays a link with an object with path prop', () => {
    const { el } = factory(
      START_LOCATION_NORMALIZED,
      { to: { path: locations.basic.string } },
      locations.basic.normalized
    )
    expect(el.querySelector('a')!.getAttribute('href')).toBe('/home')
  })

  it('can be active', () => {
    const { el } = factory(
      locations.basic.normalized,
      { to: locations.basic.string },
      locations.basic.normalized
    )
    expect(el.querySelector('a')!.className).toContain('router-link-active')
  })

  it('is not active with more repeated params', () => {
    const { el } = factory(
      locations.repeatedParams2.normalized,
      { to: locations.repeatedParams3.string },
      locations.repeatedParams3.normalized
    )
    expect(el.querySelector('a')!.className).toBe('')
  })

  it('is not active with partial repeated params', () => {
    const { el } = factory(
      locations.repeatedParams3.normalized,
      { to: locations.repeatedParams2.string },
      locations.repeatedParams2.normalized
    )
    expect(el.querySelector('a')!.className).toBe('')
  })

  it.todo('can be active as an alias')
  it.todo('can be exact-active as an alias')
  it.todo('is active when a child is active')
  it.todo('only the children is exact-active')
  it.todo('is not active if the parent is active')

  it('can be exact-active', () => {
    const { el } = factory(
      locations.basic.normalized,
      { to: locations.basic.string },
      locations.basic.normalized
    )
    expect(el.querySelector('a')!.className).toContain(
      'router-link-exact-active'
    )
  })

  it('calls ensureLocation', () => {
    const { router } = factory(
      START_LOCATION_NORMALIZED,
      { to: locations.basic.string },
      locations.basic.normalized
    )
    expect(router.resolve).toHaveBeenCalledTimes(1)
    expect(router.resolve).toHaveBeenCalledWith(locations.basic.string)
  })

  // TODO: call when we can test this
  it.skip('calls router.push when clicked', async () => {
    const { router, el } = factory(
      START_LOCATION_NORMALIZED,
      { to: locations.basic.string },
      locations.basic.normalized
    )
    el.click()
    await tick()
    expect(router.push).toHaveBeenCalledTimes(1)
    expect(router.push).toHaveBeenCalledWith(locations.basic.normalized)
  })

  describe('v-slot', () => {
    function factory(
      currentLocation: RouteLocationNormalized,
      propsData: any,
      resolvedLocation: RouteLocationNormalized
    ) {
      const router = {
        history: createMemoryHistory(),
        createHref(to: RouteLocationNormalized): string {
          return this.history.base + to.fullPath
        },
        resolve: jest.fn(),
        push: jest.fn().mockResolvedValue(resolvedLocation),
        currentRoute: ref(markNonReactive(currentLocation)),
        setActiveApp: jest.fn(),
      }

      router.resolve.mockReturnValueOnce(resolvedLocation)
      const { app, el } = mount(router as any, {
        template: `<RouterLink :to="to" v-slot="data">
        route: {{ JSON.stringify(data.route) }}
        href: "{{ data.href }}"
        isActive: "{{ data.isActive }}"
        isExactActive: "{{ data.isExactActive }}"
      </RouterLink>`,
        components: { RouterLink } as any,
        setup() {
          const to = ref(propsData.to)

          return { to }
        },
      })

      return { app, router, el }
    }

    it('provides information on v-slot', () => {
      const { el } = factory(
        locations.basic.normalized,
        { to: locations.basic.string },
        locations.basic.normalized
      )

      expect(el.innerHTML).toMatchSnapshot()
    })
  })
})
