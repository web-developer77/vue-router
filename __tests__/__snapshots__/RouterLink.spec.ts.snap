// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouterLink v-slot provides information on v-slot 1`] = `"<a class=\\"router-link-active\\" href=\\"/home\\"> route: {\\"fullPath\\":\\"/home\\",\\"path\\":\\"/home\\",\\"params\\":{},\\"meta\\":{},\\"query\\":{},\\"hash\\":\\"\\",\\"matched\\":[]} href: \\"/home\\" isActive: \\"true\\" </a>"`;
